### PR TITLE
[Storage] Use structs for model types where appropriate

### PR DESCRIPTION
**Pull request by ShivangiReja**
Changing the code generation to use structs for model types identified in the [API review](https://apiview.azurewebsites.net/Assemblies/Review/c52f059a-6d57-4d40-b7d2-a5816565a660).

This PR changes the `AccountInfo`, `PageInfo` and `PageRange` models to use struct as per the [guidelines ](https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/struct).

Here is the gist showing how the generated code would look like: https://gist.github.com/ShivangiReja/ccb498e4fba96452be7a52b156d17966

### CODE REVIEW

File: sdk/storage/Azure.Storage.Blobs/src/Generated/BlobRestClient.cs [hunk 1]
```diff
@@ -7814,36 +7817,43 @@ public static partial class PageBlob
                 {
                     case 201:
                     {
-                        // Create the result
-                        Azure.Storage.Blobs.Models.PageInfo _value = new Azure.Storage.Blobs.Models.PageInfo();
 
                         // Get response headers
                         string _header;
+                        Azure.Core.Http.ETag eTag = default;
```
Comment (pakrym) [inline sdk/storage/Azure.Storage.Blobs/src/Generated/BlobRestClient.cs]: Other variables are defined with `_` prefix. While I like how they look without prefix we should be consistent.

File: sdk/storage/Azure.Storage.Blobs/src/Generated/BlobRestClient.cs [hunk 2]
```diff
@@ -7814,36 +7817,43 @@ public static partial class PageBlob
...
-                            _value.BlobSequenceNumber = long.Parse(_header, System.Globalization.CultureInfo.InvariantCulture);
+                            blobSequenceNumber = long.Parse(_header, System.Globalization.CultureInfo.InvariantCulture);
                         }
                         if (response.Headers.TryGetValue("x-ms-encryption-key-sha256", out _header))
                         {
-                            _value.EncryptionKeySha256 = _header;
+                            encryptionKeySha256 = _header;
                         }
 
+                        // Create the result
```
Comment (pakrym) [inline sdk/storage/Azure.Storage.Blobs/src/Generated/BlobRestClient.cs]: I would remove this comment, it's not very useful to the reader.

File: sdk/storage/Azure.Storage.Blobs/src/Generated/BlobRestClient.cs [hunk 3]
```diff
@@ -12782,29 +12810,67 @@ public partial struct AccessTier : System.IEquatable<AccessTier>
...
+        /// Check if two AccountInfo instances are equal.
+        /// </summary>
+        /// <param name="obj">The instance to compare to.</param>
+        /// <returns>True if they're equal, false otherwise.</returns>
+        public override bool Equals(object obj) => obj is AccountInfo && Equals((AccountInfo)obj);
+
+        /// <summary>
+        /// Get a hash code for the AccountInfo.
+        /// </summary>
+        public override int GetHashCode(){
```
Comment (pakrym) [inline sdk/storage/Azure.Storage.Blobs/src/Generated/BlobRestClient.cs]: `{` on a new line.

File: sdk/storage/Azure.Storage.Blobs/src/Generated/BlobRestClient.cs [hunk 4]
```diff
@@ -18025,53 +18087,111 @@ public static partial class BlobsModelFactory
...
+        /// <returns>True if they're equal, false otherwise.</returns>
+        public bool Equals(PageInfo other)
+        {
+            if (!ETag.Equals(other.ETag))
+                return false;
+            if (!LastModified.Equals(other.LastModified))
+                return false;
+            if (!ContentHash.Equals(other.ContentHash))
+                return false;
+            if (!XMSContentCrc64.Equals(other.XMSContentCrc64))
```
Comment (pakrym) [inline sdk/storage/Azure.Storage.Blobs/src/Generated/BlobRestClient.cs]: Can `XMSContentCrc64` or other reference type properties be null? If the can we should guard for it.

File: sdk/storage/Azure.Storage.Blobs/src/Generated/BlobRestClient.cs [hunk 5]
```diff
@@ -18025,53 +18087,111 @@ public static partial class BlobsModelFactory
...
+        /// </summary>
+        public override int GetHashCode(){
+            var hashCode = new Azure.Core.HashCodeBuilder();
+            hashCode.Add(ETag);
+            hashCode.Add(LastModified);
+            hashCode.Add(ContentHash);
+            hashCode.Add(XMSContentCrc64);
+            hashCode.Add(BlobSequenceNumber);
+            hashCode.Add(EncryptionKeySha256);
+
```
Comment (pakrym) [inline sdk/storage/Azure.Storage.Blobs/src/Generated/BlobRestClient.cs]: I would remove this newline.

File: sdk/storage/Azure.Storage.Blobs/src/Generated/BlobRestClient.cs [hunk 6]
```diff
@@ -17843,53 +17905,111 @@ public static partial class BlobsModelFactory
...
+        /// Get a hash code for the PageInfo.
+        /// </summary>
+        public override int GetHashCode(){
+            var hashCode = new Azure.Core.HashCodeBuilder();
+            hashCode.Add(ETag);
+            hashCode.Add(LastModified);
+            hashCode.Add(ContentHash);
+            hashCode.Add(XMSContentCrc64);
+            hashCode.Add(BlobSequenceNumber);
+            hashCode.Add(EncryptionKeySha256);
```
Comment (pakrym) [inline sdk/storage/Azure.Storage.Blobs/src/Generated/BlobRestClient.cs]: If comparer is used in Equals the same comparer should be passed into `hashCode.Add` so the comparisons are consistent.

File: sdk/storage/generate.ps1 [hunk 7]
```diff
@@ -10,7 +10,7 @@ autorest --use=$PSScriptRoot/Azure.Storage.Common/swagger/Generator/ --verbose
 cd $PSScriptRoot/Azure.Storage.Queues/swagger/
 autorest --use=$PSScriptRoot/Azure.Storage.Common/swagger/Generator/ --verbose
 
-cd $PSScriptRoot/Azure.Storage.Files.DataLake/swagger/
-autorest --use=$PSScriptRoot/Azure.Storage.Common/swagger/Generator/ --verbose
+# cd $PSScriptRoot/Azure.Storage.Files.DataLake/swagger/
```
Comment (pakrym) [inline sdk/storage/generate.ps1]: revert?

**Comment by pakrym** [comment]
Use the Ordinal/OrginalIgnoreCase comparer for strings.

**Comment by pakrym** [comment]
Figure out if we need to compare byte arrays by value

File: sdk/storage/Azure.Storage.Blobs/src/Generated/BlobRestClient.cs [hunk 8]
```diff
@@ -17843,53 +17906,127 @@ public static partial class BlobsModelFactory
...
+        /// </summary>
+        public override int GetHashCode()
+        {
+            var hashCode = new Azure.Core.HashCodeBuilder();
+            hashCode.Add(ETag);
+            hashCode.Add(LastModified);
+            hashCode.Add(ContentHash);
+            hashCode.Add(XMSContentCrc64);
+            hashCode.Add(BlobSequenceNumber);
+            if (EncryptionKeySha256 != null)
```
Comment (AlexanderSher) [inline sdk/storage/Azure.Storage.Blobs/src/Generated/BlobRestClient.cs]: `XMSContentCrc64` and `ContentHash` can be `null` as well.

File: sdk/storage/Azure.Storage.Blobs/tests/PageInfoTest.cs [hunk 9]
```diff
@@ -0,0 +1,178 @@
...
+using System;
+using Azure.Storage.Blobs.Models;
+using NUnit.Framework;
+
+namespace Azure.Storage.Blobs.Tests
+{
+    public class PageInfoTest
+    {
+        [Test]
+        public void EqualsReturnsTrueForEqualValues()
```
Comment (AlexanderSher) [inline sdk/storage/Azure.Storage.Blobs/tests/PageInfoTest.cs]: Are these tests are code-generated?

File: sdk/storage/Azure.Storage.Blobs/src/Generated/BlobRestClient.cs (hunk 8 above)
Comment (ShivangiReja) [reply sdk/storage/Azure.Storage.Blobs/src/Generated/BlobRestClient.cs]: If `XMSContentCrc64` or `ContentHash` is null then [Add(value)](https://github.com/Azure/azure-sdk-for-net/blob/d3d19ff0057bb3443cf39929b8e79d3ba208bd85/sdk/core/Azure.Core/src/Shared/HashCodeBuilder.cs#L262) method will set the value 0.

File: sdk/storage/Azure.Storage.Blobs/tests/PageInfoTest.cs (hunk 9 above)
Comment (ShivangiReja) [reply sdk/storage/Azure.Storage.Blobs/tests/PageInfoTest.cs]: No, these are hand written tests.

File: sdk/storage/Azure.Storage.Blobs/src/Generated/BlobRestClient.cs (hunk 8 above)
Comment (ShivangiReja) [reply sdk/storage/Azure.Storage.Blobs/src/Generated/BlobRestClient.cs]: I've also written a test case to verify the [null case scenarios](https://github.com/Azure/azure-sdk-for-net/blob/a99fe024bbef0e5771013a358920063821f8bc6b/sdk/storage/Azure.Storage.Blobs/tests/PageInfoTest.cs#L42).

Comment (ShivangiReja) [reply sdk/storage/Azure.Storage.Blobs/src/Generated/BlobRestClient.cs]: I've a test case to verify the [null case scenarios](https://github.com/Azure/azure-sdk-for-net/blob/a99fe024bbef0e5771013a358920063821f8bc6b/sdk/storage/Azure.Storage.Blobs/tests/PageInfoTest.cs#L42).

**Comment by ShivangiReja** [comment]
@tg-msft @pakrym Using [StructuralComparisons.StructuralEquality](https://docs.microsoft.com/en-us/dotnet/api/system.collections.structuralcomparisons.structuralequalitycomparer?view=netcore-3.0) for array comparison and HashCode generation.


We cannot use `System.Linq.Enumerable.SequenceEqual` because it is not compatible with HashCode generation logic we were using before. It was using the default `Object.GetHashCode()` method which generates HashCode using the memory reference whereas `System.Linq.Enumerable.SequenceEqual` uses structural equality for comparison.

**Comment by pakrym** [comment]
> @tg-msft @pakrym Using StructuralComparisons.StructuralEquality for array comparison and HashCode generation.

This is way better than `SequenceEqual` good job finding it!

File: sdk/storage/Azure.Storage.Blobs/src/Generated/BlobRestClient.cs (hunk 1 above)
Comment (tg-msft) [reply sdk/storage/Azure.Storage.Blobs/src/Generated/BlobRestClient.cs]: I'm not actually sure if we want that here.  I prefixed the "well known variables" with an `_` to avoid conflicts, but these are the sort of values that could cause conflicts with the `_` variables.  For example, if somebody adds a struct with a property called `xml`, we wouldn't want that to conflict with our "well known variable" `_xml`.  I think this should probably change back what you originally had.

File: sdk/storage/Azure.Storage.Common/swagger/Generator/src/generator.ts [hunk 10]
```diff
@@ -617,6 +626,17 @@ function generateOperation(w: IndentWriter, serviceModel: IServiceModel, group:
                     });
                 }
             }
+            if (response.struct) {
+                w.line();
+                const Separator = IndentWriter.createFenceposter();
```
Comment (tg-msft) [inline sdk/storage/Azure.Storage.Common/swagger/Generator/src/generator.ts]: nit - don't we case locals as `separator` in TypeScript?

File: sdk/storage/Azure.Storage.Common/swagger/Generator/src/generator.ts [hunk 11]
```diff
@@ -913,7 +938,78 @@ function generateObject(w: IndentWriter, model: IServiceModel, type: IObjectType
...
+                        w.line();
+                        w.line(`return true;`);
+                    });
+                    w.line();
+                    w.line(`/// <summary>`);
+                    w.line(`/// Check if two ${naming.type(type.name)} instances are equal.`);
+                    w.line(`/// </summary>`);
+                    w.line(`/// <param name="obj">The instance to compare to.</param>`);
+                    w.line(`/// <returns>True if they're equal, false otherwise.</returns>`);
+                    w.line(`public override bool Equals(object obj) => obj is ${naming.type(type.name)} && Equals((${naming.type(type.name)})obj);`);
```
Comment (tg-msft) [inline sdk/storage/Azure.Storage.Common/swagger/Generator/src/generator.ts]: @pakrym - should we add an `[EditorBrowsable(false)]` here?

File: sdk/storage/Azure.Storage.Common/swagger/Generator/src/generator.ts [hunk 12]
```diff
@@ -913,7 +938,78 @@ function generateObject(w: IndentWriter, model: IServiceModel, type: IObjectType
...
+                    w.line();
+                    w.line(`/// <summary>`)
+                    w.line(`/// Get a hash code for the ${naming.type(type.name)}.`);
+                    w.line(`/// </summary>`);
+                    w.line(`public override int GetHashCode()`);
```
Comment (tg-msft) [inline sdk/storage/Azure.Storage.Common/swagger/Generator/src/generator.ts]: We should definitely add `[EditorBrowsable(false)]` here.

File: sdk/storage/Azure.Storage.Common/swagger/Generator/src/generator.ts [hunk 13]
```diff
@@ -913,7 +938,78 @@ function generateObject(w: IndentWriter, model: IServiceModel, type: IObjectType
...
+                    w.scope('{', '}', () => {
+                        w.line(`var hashCode = new Azure.Core.HashCodeBuilder();`);
+                        for (const property of properties) {
+                            if (types.getDeclarationType(property.model, property.required, property.readonly) === "string") {
+                                w.line(`if (EncryptionKeySha256 != null)`)
```
Comment (tg-msft) [inline sdk/storage/Azure.Storage.Common/swagger/Generator/src/generator.ts]: Should this be the property name instead of hardcoded `EncryptionKeySha256 `?

File: sdk/storage/Azure.Storage.Common/swagger/Generator/src/generator.ts [hunk 14]
```diff
@@ -1202,6 +1312,16 @@ function generateDeserialize(w: IndentWriter, service: IService, type: IObjectTy
                 }
             }
         }
+        if (type.struct) {
+            const Separator = IndentWriter.createFenceposter();
```
Comment (tg-msft) [inline sdk/storage/Azure.Storage.Common/swagger/Generator/src/generator.ts]: Nit - casing

File: sdk/storage/Azure.Storage.Common/swagger/Generator/src/models.ts [hunk 15]
```diff
@@ -237,7 +238,8 @@ export interface IResponse {
     model?: IModelType,
     exception?: boolean,
     public: boolean,
-    returnStream?: boolean
+    returnStream?: boolean,
+    struct: boolean
```
Comment (tg-msft) [inline sdk/storage/Azure.Storage.Common/swagger/Generator/src/models.ts]: Does the response need a `struct` property?  Can't you just set it on `model` in `createResponse`?

**Review by tg-msft — APPROVED** [review]
Awesome work Shivangi!

File: sdk/storage/Azure.Storage.Common/swagger/Generator/src/generator.ts (hunk 11 above)
Comment (pakrym) [reply sdk/storage/Azure.Storage.Common/swagger/Generator/src/generator.ts]: Yep

File: sdk/storage/Azure.Storage.Blobs/tests/PageInfoTest.cs [hunk 16]
```diff
@@ -0,0 +1,205 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using Azure.Storage.Blobs.Models;
+using NUnit.Framework;
+
+namespace Azure.Storage.Blobs.Tests
+{
+    public class PageInfoTest
```
[the remark this answers is not in the document]
Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Blobs/tests/PageInfoTest.cs]: I agree, I wasn't sure what this test was for at first.

**Review by seanmcc-msft — APPROVED** [review]

File: sdk/storage/Azure.Storage.Common/src/Azure.Storage.Common.csproj [hunk 17]
```diff
@@ -24,5 +24,6 @@
     -->
     <Compile Include="$(AzureCoreSharedSources)ArrayBufferWriter.cs" Link="AzureCoreShared\%(RecursiveDir)\%(Filename)%(Extension)" />
     <Compile Include="$(AzureCoreSharedSources)ForwardsClientCallsAttribute.cs" Link="AzureCoreShared\%(RecursiveDir)\%(Filename)%(Extension)" />
+    <Compile Include="$(AzureCoreSharedSources)HashCodeBuilder.cs" />
```
[the remark this answers is not in the document]
Comment (ShivangiReja) [reply sdk/storage/Azure.Storage.Common/src/Azure.Storage.Common.csproj]: It was working as expected by using ` <Compile Include="$(AzureCoreSharedSources)HashCodeBuilder.cs" />` only, that's why I didn't realize about adding `Link="AzureCoreShared\%(RecursiveDir)\%(Filename)%(Extension)"`.

I'll update it :)

File: sdk/storage/Azure.Storage.Common/swagger/Generator/src/generator.ts (hunk 13 above)
Comment (ShivangiReja) [reply sdk/storage/Azure.Storage.Common/swagger/Generator/src/generator.ts]: Good catch :) Thank you!

File: sdk/storage/Azure.Storage.Common/swagger/Generator/src/models.ts (hunk 15 above)
Comment (ShivangiReja) [reply sdk/storage/Azure.Storage.Common/swagger/Generator/src/models.ts]: > Can't you just set it on model in `createResponse`?

How will we set `struct` in [createResponseType](https://github.com/Azure/azure-sdk-for-net/blob/9ca9e5b302baeb2556013389c935fc034d005ee5/sdk/storage/Azure.Storage.Common/swagger/Generator/src/serviceModel.ts#L963)?

File: sdk/storage/Azure.Storage.Common/swagger/Generator/src/generator.ts (hunk 10 above)
Comment (ShivangiReja) [reply sdk/storage/Azure.Storage.Common/swagger/Generator/src/generator.ts]: Yes, I got confused between Typescript and .Net. Will update it.

**Review by amnguye — APPROVED** [review]
